### PR TITLE
Update location handling

### PR DESCRIPTION
**Pull request by jwalgran**
There were multiple problems with the previous map view controller
location handling implementation.

- If the accuracy of location results is never less than 100m the
  location is never shown on the map.
- The spinner on the location button never stops if there is a failure
  to find the location or the accuracy of the locations is never less
  than 100m.
- OTMLocationManager was using a deprecated API.
- Location does not update continuously on the map view.
- The app uses multiple `CLLocationManager` instances, but that API is
  designed to have 1 manager per app.

I have addressed most of these issues by:

- Adding a `CLLocationManager` instance to the app delegate.
- Removing references to `OTMLocationManager` from the `OTMMapViewController`, `OTMNearbyTreesViewController`, and `OTMInstanceSelectTableViewController` and making each of those controllers a `CLLocationManagerDelegate`.
- Setting `showsUserLocation` on the `MKMapView` to continuously display location on the map using the built-in MapKit functionality.
- Renaming the actions triggered by the location button on the map view and changing their functionality to control zooming of the map, not starting and stopping location updates, which now run continuously.

All view controllers that depend on location new use the shared `CLLocationManager` on the app delegate, allowing us to remove `OTMLocationManager` entirely.

I have opted to eventually remove `OTMLocationManager` because:

- Each of the three view controllers that need to know the user's location have different requirements.
- Building an abstraction that served the needs of the three different views requires maintaining a API that replicates parts of the built-in delegate behavior.
- The view controller-as-delegate pattern is a strong iOS development convention.

I added the CLLocation+AgeInSecondsLessThan category in order to share the filtering logic with multiple `CLLocationManager` delegate classes.

In the nearby trees view I increased the accuracy and decreased minimum distance traveled before update to improve the responsiveness of results displayed in the table view, which I discovered during some walk-around-outside testing did  not update frequently enough.

My initial implementation had each of the 3 location aware view controllers using `viewWillDisappear:` to shut off location updates but I discovered in both in testing and in researching the API that there is no guarantee that `viewWillDisappear:` in one view controller will run before `viewWillAppear:` in the other. For this reason, I added calls to disable location updates to `viewWillAppear:` in view controllers that do not need location updates.

---

##### Testing

I created a test instance on production centered near my office location and tested location tracking by walking around outside. 

I used these  [three 3 GPX files ](https://github.com/OpenTreeMap/otm-ios/files/522281/New.York.7th.Ave.GPX.zip) for additional testing of location updates in the simulator. They are locations along 7th Ave in New York City south of the default New York location built into Xcode. Adding the locations to the workspace and switching between them simulates a walk down 7th ave.

- If the user is not logged in, the instance select view should show nearby instances.
- On the map view, the map dot should follow the user's location (Note the location dot is colored using the tint color and cannot be changed independently)
- If the user is located within the instance bounds, the map should zoom to the user's location the first time it appears. Scrolling the map to a different location, switching to a different view, and then returning to the map should not trigger a zoom.
- Clicking the location icon on the map view should zoom to the user's current location.
- If the user is located outside the bounds of an instance, the map should never zoom to their location.
- The nearby trees view should update as the user moves. The list should be sorted by proximity.

Here is a screenshot of the arrangement of trees I added to a test instance along 7th Ave in order to test the nearby trees view

screen shot 2016-10-11 at 9 55 33 am

---

Connects to #318 
Connects to #292

### CODE REVIEW

**Comment by RickMohr** [comment]
This appears to be a regression:
1) Launch the app and open treemap A
2) Switch to treemap B
3) Click Hardware > Home, and then return to the app (or restart the app)
Expected: display shows treemap B
Actual: display shows treemap A

**Comment by RickMohr** [comment]
It's a bit surprising that if my location is not in the current treemap bounds then hitting the "zoom to my location" button does nothing. Ideally we'd display a message so indicating. But that's outside the scope of this PR, and not a big problem.

File: OpenTreeMap/src/OTM/Controllers/OTMNearbyTreesViewController.m [hunk 1]
```diff
@@ -74,38 +75,36 @@ - (void)viewDidLoad
...
-    } else {
-        [self.locationManager startUpdatingLocation];
+    if ([CLLocationManager locationServicesEnabled]) {
+        // Required to get iOS8 location services to run.
+        if ([[SharedAppDelegate locationManager] respondsToSelector:@selector(requestWhenInUseAuthorization)]) {
+            [[SharedAppDelegate locationManager] requestWhenInUseAuthorization];
+            [self startUpdatingLocation];
+        } else {
+            [self startUpdatingLocation];
+        }
```
Comment (RickMohr) [inline OpenTreeMap/src/OTM/Controllers/OTMNearbyTreesViewController.m]: Similarly could delete this `else`

**Comment by jwalgran** [comment]
I was able to reproduce the `EXC_BAD_ACCESS` problem on the `master` branch. I have a proposed fix for it that I will open as a separate PR.

**Comment by jwalgran** [comment]
I created https://github.com/OpenTreeMap/otm-ios/issues/324 to cover notifying the user that the app is not zooming to their location because they are too far from the center of the instance.

**Comment by jwalgran** [comment]
@RickMohr I made the small changes you suggested and created issues for out-of-scope items.

**Comment by RickMohr** [comment]
+1

**Pull request merged**